stats/opentelemetry/client_metrics.go

```diff
@@ -0,0 +1,222 @@
+
+// TagRPC implements per RPC attempt context management.
+func (csh *clientStatsHandler) TagRPC(ctx context.Context, info *stats.RPCTagInfo) context.Context {
+	mi := &metricsInfo{ // populates information about RPC start.
+		startTime: time.Now(),
+	}
+	ri := &rpcInfo{
+		mi: mi,
+	}
+	return setRPCInfo(ctx, ri)
+}
+
+func (csh *clientStatsHandler) HandleRPC(ctx context.Context, rs stats.RPCStats) {
+	ri := getRPCInfo(ctx)
+	if ri == nil {
+		// Shouldn't happen because TagRPC populates this information.
+		return
+	}
+	csh.processRPCEvent(ctx, rs, ri.mi)
+}
+
+func (csh *clientStatsHandler) processRPCEvent(ctx context.Context, s stats.RPCStats, mi *metricsInfo) {
+	switch st := s.(type) {
+	case *stats.InHeader, *stats.OutHeader, *stats.InTrailer, *stats.OutTrailer:
+	case *stats.Begin:
+		ci := getCallInfo(ctx)
+		if ci == nil {
+			logger.Error("ctx passed into client side stats handler metrics event handling has no metrics data present")
+			return
+		}
+
+		if csh.clientMetrics.attemptStarted != nil {
+			csh.clientMetrics.attemptStarted.Add(ctx, 1, metric.WithAttributes(attribute.String("grpc.method", ci.method), attribute.String("grpc.target", ci.target)))
+		}
+	case *stats.OutPayload:
+		atomic.AddInt64(&mi.sentCompressedBytes, int64(st.CompressedLength))
+	case *stats.InPayload:
+		atomic.AddInt64(&mi.recvCompressedBytes, int64(st.CompressedLength))
+	case *stats.End:
+		csh.processRPCEnd(ctx, mi, st)
+	default:
+		// Shouldn't happen. gRPC calls into stats handler, and will never not
+		// be one of the types above.
+		logger.Errorf("Received unexpected stats type (%T) with data: %v", s, s)
+	}
+}
+
+func (csh *clientStatsHandler) processRPCEnd(ctx context.Context, mi *metricsInfo, e *stats.End) {
+	ci := getCallInfo(ctx)
+	if ci == nil {
+		// Shouldn't happen, set by interceptor, defensive programming.
+		return
+	}
+	latency := float64(time.Since(mi.startTime)) / float64(time.Second)
+	var st string
+	if e.Error != nil {
+		s, _ := status.FromError(e.Error)
+		st = canonicalString(s.Code())
+	} else {
+		st = "OK"
+	}
+	clientAttributeOption := metric.WithAttributes(attribute.String("grpc.method", ci.method), attribute.String("grpc.target", ci.target), attribute.String("grpc.status", st))
+	if csh.clientMetrics.attemptDuration != nil {
+		csh.clientMetrics.attemptDuration.Record(ctx, latency, clientAttributeOption)
+	}
+
+	if csh.clientMetrics.attemptSentTotalCompressedMessageSize != nil {
+		csh.clientMetrics.attemptSentTotalCompressedMessageSize.Record(ctx, atomic.LoadInt64(&mi.sentCompressedBytes), clientAttributeOption)
+	}
+
+	if csh.clientMetrics.attemptRcvdTotalCompressedMessageSize != nil {
+		csh.clientMetrics.attemptRcvdTotalCompressedMessageSize.Record(ctx, atomic.LoadInt64(&mi.recvCompressedBytes), clientAttributeOption)
+	}
+}
+
+const (
+	// ClientAttemptStartedName is the name of the client attempt started
+	// metric.
+	ClientAttemptStartedName = MetricName("grpc.client.attempt.started")
```

Comment:
s/Name// please.  Note that several of your defined metrics don't have the `Name` suffix already.